internal/controller/supervisorconfig/upstreamwatchers/upstream_watchers.go

```diff
@@ -183,16 +183,18 @@ func HasPreviousSuccessfulTLSConnectionConditionForCurrentSpecGenerationAndSecre
 	return false
 }
 
-func HasPreviousSuccessfulSearchBaseConditionForCurrentGeneration(secretVersionCache *SecretVersionCache, currentGeneration int64, upstreamStatusConditions []v1alpha1.Condition, upstreamName string, currentSecretVersion string, config *upstreamldap.ProviderConfig) bool {
+func HasPreviousSuccessfulSearchBaseConditionForCurrentGeneration(secretVersionCache *SecretVersionCache, currentGeneration int64, upstreamStatusConditions []v1alpha1.Condition, upstreamName string, currentSecretVersion string, previouslyValidatedSecretVersion string, config *upstreamldap.ProviderConfig) bool {
```

Comment:
Please refactor `SecretVersionCache` into an interface that looks something like:

```go
type ValidatedCache interface {
    Get(upstreamName, resourceVersion string) (ValidatedSettings, bool)
    Set(upstreamName, resourceVersion string, settings ValidatedSettings)
}
```

IMO this code needs a larger refactor where it only considers RV and generation together (but we can defer that for now):

```go
type ValidatedCache interface {
    Get(upstreamName, resourceVersion string, generation int64) (ValidatedSettings, bool)
    Set(upstreamName, resourceVersion string, generation int64, settings ValidatedSettings)
}
```

The code has a lot of complexity due to how it tries to cache RV and generation as distinct bits of the cache, but that is a pointless optimization as both RV and generation should almost never change.  Thus writing the simpler code here is more appropriate.